translator/internaldata/oc_to_metrics.go

```diff
@@ -0,0 +1,322 @@
+}
+
+func ocMetricToResourceMetrics(ocMetric *ocmetrics.Metric, node *occommon.Node, out data.ResourceMetrics) {
+	ocNodeResourceToInternal(node, ocMetric.Resource, out.Resource())
+	ilms := out.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	metrics := ilms.At(0).Metrics()
+	metrics.Resize(1)
+	ocMetricToInternal(ocMetric, metrics.At(0))
+}
+
+// ocMetricToInternal conversts ocMetric to internal representation and fill metric
+func ocMetricToInternal(ocMetric *ocmetrics.Metric, metric data.Metric) {
+	metricDescriptor := metric.MetricDescriptor()
+	metricDescriptor.InitEmpty()
+	descriptorToInternal(ocMetric.GetMetricDescriptor(), metricDescriptor)
+	labels := getInternalLabels(ocMetric)
+	setDataPoints(ocMetric, metric, labels)
+}
+
+func descriptorToInternal(ocDescriptor *ocmetrics.MetricDescriptor, descriptor data.MetricDescriptor) {
+	if ocDescriptor == nil {
+		return
+	}
+
+	descriptorType := descriptorTypeToInternal(ocDescriptor.Type)
+	if descriptorType == invalidMetricType {
+		return
+	}
+
+	descriptor.SetType(descriptorType)
+	descriptor.SetDescription(ocDescriptor.GetDescription())
+	descriptor.SetName(ocDescriptor.GetName())
+	descriptor.SetUnit(ocDescriptor.GetUnit())
+}
+
+func descriptorTypeToInternal(t ocmetrics.MetricDescriptor_Type) data.MetricType {
+	switch t {
+	case ocmetrics.MetricDescriptor_UNSPECIFIED:
+		return data.MetricTypeUnspecified
+	case ocmetrics.MetricDescriptor_GAUGE_INT64:
+		return data.MetricTypeGaugeInt64
+	case ocmetrics.MetricDescriptor_GAUGE_DOUBLE:
+		return data.MetricTypeGaugeDouble
+	case ocmetrics.MetricDescriptor_GAUGE_DISTRIBUTION:
+		return data.MetricTypeGaugeHistogram
+	case ocmetrics.MetricDescriptor_CUMULATIVE_INT64:
+		return data.MetricTypeCounterInt64
+	case ocmetrics.MetricDescriptor_CUMULATIVE_DOUBLE:
+		return data.MetricTypeCounterDouble
+	case ocmetrics.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		return data.MetricTypeCumulativeHistogram
+	case ocmetrics.MetricDescriptor_SUMMARY:
+		return data.MetricTypeSummary
+	default:
+		return invalidMetricType
+	}
+}
+
+// getInternalLabels scans OC metric labels and returns OC timeseries idx -> labels map.
+func getInternalLabels(ocMetric *ocmetrics.Metric) []data.StringMap {
+	ocLabelsKeys := ocMetric.GetMetricDescriptor().GetLabelKeys()
+	labelsCount := len(ocLabelsKeys)
+	timeseriesCount := len(ocMetric.GetTimeseries())
+	if timeseriesCount == 0 {
+		return nil
+	}
+
+	labelsByTimeseriesIdx := make([]data.StringMap, timeseriesCount)
+
+	// Scan timeseries and fill the OC timeseries idx -> internal labels map
+	for i := 0; i < timeseriesCount; i++ {
+		ts := ocMetric.GetTimeseries()[i]
+		labelsByTimeseriesIdx[i] = data.NewEmptyStringMap(labelsCount)
+		for l := 0; l < labelsCount; l++ {
+			ocLabelValues := ts.GetLabelValues()
+			if ocLabelValues == nil || !ocLabelValues[l].GetHasValue() {
+				continue
+			}
+			labelsByTimeseriesIdx[i].Insert(ocLabelsKeys[l].Key, ocLabelValues[l].Value)
+		}
+	}
+
+	return labelsByTimeseriesIdx
+}
```

Comment:
Can you explain why is it better to pre-generate all the label sets vs generating for every point?

Reply:
My intention was to reduce amount of work that needs to be done to prepare the labels. After giving it a second thought, it looks like a wrong approach...

OC has one set of labels per timeseries, and each timeseries has more than data point, OTLP/internal has one set per datapoint. Here we pre-generate a map of labels per OC timeseries. It gives us ability to reuse one set of labels for many datapoints, which is wrong. I either need to make a copy for every other usage of a labels set or generate it every time for every datapoint.

Reply:
In practice would not be an error because I've almost never seen multiple points in opencensus that's why we removed that optimization, but there may be some corner cases when that happens.